builder/test/stencil_pipeline.cc

```diff
@@ -0,0 +1,157 @@
+#include <gmock/gmock.h>
+#include <gtest/gtest.h>
+
+#include <functional>
+#include <vector>
+
+#include "builder/pipeline.h"
+#include "builder/test/context.h"
+#include "builder/test/funcs.h"
+#include "builder/test/util.h"
+#include "runtime/expr.h"
+#include "runtime/pipeline.h"
+
+namespace slinky {
+
+class stencil : public testing::TestWithParam<std::tuple<bool, int, int, int>> {};
+
+INSTANTIATE_TEST_SUITE_P(stride_dilation_size, stencil,
+    testing::Combine(testing::Bool(), testing::Values(1, 2, 3), testing::Values(1, 2, 3), testing::Values(1, 2, 3)),
+    test_params_to_string<stencil::ParamType>);
+
+TEST_P(stencil, 1d) {
+  const bool no_alias_buffers = std::get<0>(GetParam());
+  const int S = std::get<1>(GetParam());
+  const int D = std::get<2>(GetParam());
+  const int K = std::get<3>(GetParam());
+
+  // Make the pipeline
+  node_context ctx;
+
+  auto in = buffer_expr::make(ctx, "in", 1, sizeof(short));
+  auto out = buffer_expr::make(ctx, "out", 1, sizeof(short));
+
+  in->dim(0).fold_factor = dim::unfolded;
+
+  auto stencil = buffer_expr::make(ctx, "stencil", 2, sizeof(short));
+
+  var x(ctx, "x");
+  var dx(ctx, "dx");
+
+  // This test computes the following stencil operation:
+  //
+  //  for i in [0, N):
+  //    for k in [0, K):
+  //      out[i] += in[i * S + k * D]
+  //
+  // Using the following approach;
+  //
+  // 1. Make a copy of the input such that stencil(x, dx) = in(x * S + dx * D)
```

Comment:
This is a very powerful idea and this comment explains well what it does, but I think it might be useful to have some at least high-level explanation why this might be useful.